working/macros/feature-specification.md

```diff
@@ -393,6 +393,115 @@ be constructed, but where you need to build up something in smaller pieces you
+include resources outside of the `lib` directory of a package - even for
+package dependencies - depending on how the package config file is configured.
+
+If the URI points to a symlink it must be followed and the final physical file
+location checked to be a valid path under a package root. Otherwise they could
+be used to circumvent this check and load a resource that is outside the scope
+of the program. **TODO**: Evaluate whether this restriction is problematic for
+any current compilation strategies, such as in bazel, and if so consider
+alternatives.
+
+Resources are read via a [Uri][]. This may be a `package:` URI, or an absolute
+URI of any other form as long as it exists under the root URI of some package
+listed in the package config.
+
+It is also intuitive for macros to accept a relative URI for resources. In
+order to support this macros should compute the absolute URI from the current
+libraries URI. This URI is accessible by introspecting on the library of the
+declaration that a macro is applied to.
+
+- **TODO**: Support for relative URIs in part files (requires a part file
+  abstraction)?
+- **TODO**: Should libraries report their fully resolved URI or the URI that
+  was used to import them? The latter would mean that files under `lib` could
+  not read resources outside of `lib`, which has both benefits and drawbacks.
+
+Lastly, since macros must return synchronously, we only expose a synchronous
+API for reading resources.
+
+The specific API is as follows, and would only be available at compile time:
+
+```dart
+/// A read-only resource API for use in macro implementation code.
+class Resource {
+  /// Either a `package:` URI, or an absolute URI which is under the root of
+  /// one or more packages in the current package config.
+  final Uri uri;
+
+  /// Creates a resource reference.
+  ///
+  /// The [uri] must be valid for this compilation, which means that it exists
+  /// under the root URI of one or more packages in the package config file.
+  ///
+  /// Throws an [InvalidResourceException] if [uri] is not valid.
+  Resource(this.uri);
+
+  /// Whether or not a resource actually exists at [uri].
+  bool get exists;
+
+  /// Synchronously reads this resource as bytes.
+  Uint8List readAsBytesSync();
+
+  /// Synchronously reads this resource as text using [encoding].
+  String readAsStringSync({Encoding encoding = utf8});
+  
+  /// Synchronously reads the resource as lines of text using [encoding].
+  List<String> readAsLinesSync({Encoding encoding = utf8});
+}
+```
+
+#### Resource Invalidation
+
+Resources that are read should be treated as source inputs to the program, and
+should invalidate the parts of the program that depended on them when they
+change.
+
+When a resource is read during compilation, it should either be cached for
+subsequent reads to use or a hash of its contents stored. No two macros should
+ever see different contents for the same resource, within the same build.
```

Comment:
Ensuring this in the worst case means storing the entire contents of the resource in memory, which could be prohibitively hard if someone decides they want their macro to operate on, say, a several-gigabyte ML model file or something.

Note that an implementation may have to cache the resource in memory *even in the common case where only one macro reads it* because it doesn't *know* that later macros won't read it until it's too late.

Reply:
Yes - that is why we don't require them to do this though. We only require the consistency aspect, which could be enforced by just storing a hash of the contents.